File: app/src/api/provenance.ts
import { USE_MOCK, request } from './http'

export type ProvenanceSource = 'ai-raw' | 'ai-edited' | 'human'

export interface ProvenanceParagraph {
  id: string
  text: string
  words: number
  source: ProvenanceSource
  runId: string | null
}

export interface SuspectedSentence {
  id: string
  paragraphId: string
  text: string
  start: number
  end: number
  source: ProvenanceSource
  score: number
  reasons: string[]
  riskRules?: SentenceRiskRule[]
}

export interface SentenceRiskRule {
  id: string
  category: string
  label: string
  score: number
  confidence: number
}

export interface ProvenanceReport {
  scope: 'chapter' | 'book'
  totalWords: number
  aiRawWords: number
  aiEditedWords: number
  humanWords: number
  paragraphs: ProvenanceParagraph[]
  suspectedSentences: SuspectedSentence[]
  sentenceRiskVersion: string
  sentenceRiskDisclaimer: string
}

export interface ProvenanceDto {
  scope: 'chapter' | 'book'
  total_words: number
  ai_raw_words: number
  ai_edited_words: number
  human_words: number
  paragraphs: Array<{
    id: string
    text: string
    words: number
    source: ProvenanceSource
    run_id: string | null
  }>
  suspected_sentences: Array<{
    id: string
    paragraph_id: string
    text: string
    start: number
    end: number
    source: ProvenanceSource
    score: number
    reasons: string[]
    risk_rules?: Array<{
      id: string
      category: string
      label: string
      score: number
      confidence: number
    }>
  }>
  sentence_risk_version: string
  sentence_risk_disclaimer: string
}

export function mapProvenanceDto(dto: ProvenanceDto): ProvenanceReport {
  return {
    scope: dto.scope,
    totalWords: dto.total_words,
    aiRawWords: dto.ai_raw_words,
    aiEditedWords: dto.ai_edited_words,
    humanWords: dto.human_words,
    paragraphs: dto.paragraphs.map((paragraph) => ({
      id: paragraph.id,
      text: paragraph.text,
      words: paragraph.words,
      source: paragraph.source,
      runId: paragraph.run_id
    })),
    suspectedSentences: dto.suspected_sentences.map((item) => ({
      id: item.id,
      paragraphId: item.paragraph_id,
      text: item.text,
      start: item.start,
      end: item.end,
      source: item.source,
      score: item.score,
      reasons: item.reasons,
      riskRules: (item.risk_rules ?? []).map((rule) => ({ ...rule }))
    })),
    sentenceRiskVersion: dto.sentence_risk_version,
    sentenceRiskDisclaimer: dto.sentence_risk_disclaimer
  }
}

export const provenanceApi = {
  async report(projectId: string, chapterId: string | null, scope: 'chapter' | 'book'): Promise<ProvenanceReport> {
    if (USE_MOCK) {
      return {
        scope,
        totalWords: 0,
        aiRawWords: 0,
        aiEditedWords: 0,
        humanWords: 0,
        paragraphs: [],
        suspectedSentences: [],
        sentenceRiskVersion: 'zh-fiction-risk-v1',
        sentenceRiskDisclaimer: '疑似句式只提示表达风险，不代表平台检测结论或 AI 鉴定。'
      }
    }
    const params = new URLSearchParams({ scope })
    if (chapterId) params.set('chapter_id', chapterId)
    return mapProvenanceDto(await request<ProvenanceDto>(`/projects/${projectId}/provenance?${params}`))
  }
}

File: server/api/provenance.py
"""Authenticated AI draft provenance reports."""

from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from api.auth import get_current_user, verify_project_access
from db.models_core import Chapter, ChapterBody, User
from db.models_usage import GenerationRun
from db.session import get_db
from services.provenance import (
    SENTENCE_RISK_VERSION,
    ProvenanceParagraph,
    classify_document,
    detect_suspected_sentences,
)

router = APIRouter()


def _summary(paragraphs: list[ProvenanceParagraph]) -> dict:
    counts = {"ai-raw": 0, "ai-edited": 0, "human": 0}
    for paragraph in paragraphs:
        counts[paragraph.source] += paragraph.words
    total = sum(counts.values())
    return {
        "total_words": total,
        "ai_raw_words": counts["ai-raw"],
        "ai_edited_words": counts["ai-edited"],
        "human_words": counts["human"],
    }


@router.get("/projects/{project_id}/provenance")
async def provenance_report(
    project_id: str,
    scope: Literal["chapter", "book"] = Query("chapter"),
    chapter_id: str | None = Query(None),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    await verify_project_access(project_id, user, db)
    statement = (
        select(Chapter, ChapterBody)
        .outerjoin(ChapterBody, ChapterBody.chapter_id == Chapter.id)
        .where(Chapter.project_id == project_id, Chapter.deleted_at.is_(None))
        .order_by(Chapter.idx)
    )
    if scope == "chapter":
        if not chapter_id:
            raise HTTPException(status_code=422, detail={"code": "CHAPTER_REQUIRED"})
        statement = statement.where(Chapter.id == chapter_id)
    rows = (await db.execute(statement)).all()
    if scope == "chapter" and not rows:
        raise HTTPException(status_code=404, detail="Chapter not found")

    known_runs = None
    if scope == "book":
        known_runs = {
            run.id: run
            for run in (await db.execute(
                select(GenerationRun).where(GenerationRun.project_id == project_id)
            )).scalars()
        }
    all_paragraphs: list[ProvenanceParagraph] = []
    for chapter, body in rows:
        if body is None:
            continue
        all_paragraphs.extend(await classify_document(
            db,
            project_id=project_id,
            chapter_id=chapter.id,
            content_json=body.content_json,
            known_runs=known_runs,
        ))
    result = {"scope": scope, **_summary(all_paragraphs)}
    result["paragraphs"] = [
        {
            "id": paragraph.paragraph_id,
            "text": paragraph.text,
            "words": paragraph.words,
            "source": paragraph.source,
            "run_id": paragraph.run_id,
        }
        for paragraph in all_paragraphs
    ] if scope == "chapter" else []
    suspected = detect_suspected_sentences(all_paragraphs) if scope == "chapter" else []
    result["suspected_sentences"] = [
        {
            "id": item.id,
            "paragraph_id": item.paragraph_id,
            "text": item.text,
            "start": item.start,
            "end": item.end,
            "source": item.source,
            "score": item.score,
            "reasons": list(item.reasons),
            "risk_rules": [
                {
                    "id": risk.rule_id,
                    "category": risk.category,
                    "label": risk.label,
                    "score": risk.score,
                    "confidence": risk.confidence,
                }
                for risk in item.risks
            ],
        }
        for item in suspected
    ]
    result["sentence_risk_version"] = SENTENCE_RISK_VERSION
    result["sentence_risk_disclaimer"] = (
        "疑似句式来自可解释的文本规则，只提示需要人工复核的表达风险，不代表平台检测结论或 AI 鉴定。"
    )
    return result
